### ISSUE TITLE
[Minimal API] Incorrect response when request body is too large

### ISSUE THREAD

**Issue by marcominerva** [dotnet/aspnetcore#47525]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I'm using .NET 7.0.4 and I'm trying to change the `MaxRequestBodySize` property in Kestrel:

    var builder = WebApplication.CreateBuilder(args);
    
    builder.WebHost.ConfigureKestrel(options =>
    {
        options.Limits.MaxRequestBodySize = 1024;
    });

With this setting, if I use a classic Controller like this:

    [ApiController]
    [Route("api/[controller]")]
    [Produces(MediaTypeNames.Application.Json)]
    public class UploadController : ControllerBase
    {
        [HttpPost("with_controller")]
        public IActionResult Upload(IFormFile file) => NoContent();
    }

When I try to upload a file larger that 1024 bytes, I obtain the following error:

    {
      "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
      "title": "One or more validation errors occurred.",
      "status": 400,
      "traceId": "00-5392ad98781087812b2bba29374a2a21-ce7ed0cde6248257-00",
      "errors": {
        "": [
          "Failed to read the request form. Request body too large. The max request body size is 1024 bytes."
        ]
      }
    }

However, if I use a Minimal API endpoint:

`app.MapPost("api/upload/with_minimalapi", (IFormFile file) => TypedResults.NoContent());`

And try to upload the same file that excedees 1024 bytes, I obtain a 200 OK response:

![image](https://user-images.githubusercontent.com/3522534/229283040-9cc9950f-a280-4014-87f7-557429378a23.png)

### Expected Behavior

The Minimal API endpoint should return a 400 Bad Request response, like the Controller.

### Steps To Reproduce

Minimal repro: https://github.com/marcominerva/UploadMaxSizeIssue

### .NET Version

7.0.202

**Comment by mitchdenny** [comment]
Thanks for reporting this @marcominerva. I dug into this a bit today to try and figure out what was going on here. There are two sets of limits that come into play when we talk about minimal APIs. The first one that you've highlighted is ```MaxRequestBodySize``` which is set via ```KestrelServerOptions```.

When minimal APIs goes to bind the ```IFormFile``` parameter it makes a request to ```HttpContext.Request.ReadFormAsync(...)``` which ultimately checks the content length against the max body here:

https://github.com/dotnet/aspnetcore/blob/0b10e1349063b1b5b39dd0435509abb52cc81d90/src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs#L30

That throws and is caught here:

https://github.com/dotnet/aspnetcore/blob/0b10e1349063b1b5b39dd0435509abb52cc81d90/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1421

I haven't had a closer look at why it doesn't result in a BadRequest yet, but the problem is in there somewhere. I also had a look at possible work arounds for this issue.

For the time being you could add something like this:

```csharp
builder.Services.Configure<FormOptions>(fo =>
{
    fo.MultipartBodyLengthLimit = 1024; // 1KB
});
```

**Comment by mitchdenny** [comment]
Actually here is where we drop the ball:

https://github.com/dotnet/aspnetcore/blob/0b10e1349063b1b5b39dd0435509abb52cc81d90/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#LL1370C25-L1370C25

You can see if reading from the form isn't successful we just return which is what causes the 200 response.

**Comment by marcominerva** [comment]
Thank you for the explanation! So, can we consider this behavior a bug?

**Comment by mitchdenny** [comment]
I think returning a 200 when an error actually occurred is definitely an issue and probably a bug. But I'm I need to find out a bit more about why it is like this. There might be something that I'm not across.

**Comment by houseofcat** [comment]
@mitchdenny @marcominerva Thanks for letting me know what I was running into :)

Opened a Design Proposal.
https://github.com/dotnet/aspnetcore/issues/47578

**Comment by mitchdenny** [comment]
@houseofcat as mentioned on the issue you opened, I closed it as a dupe of this issue (lets try and keep the conversation in one place). In terms of fixing the issue I think that there is a case for two approaches with errors. One is a 500, and the other is 400.

Generally speaking, for parameter binding errors we return a 400. But this one is a little bit more interesting because we are actually failing because of a max request size limit which is independent of parameter binding.

There is also the behavior of what happens when we exceed size limits placed on `FormOptions` - we probably want to be consistent with the error that is raised there.

The case for 400 is that its consistent with other parameter binding errors.
The case for 500 is that the error has more to do with the overall HTTP request size.
